xiaomi.vacuum.gen1/firmwarebuilder/flasher.py

```diff
@@ -15,63 +44,128 @@ def md5(fname):
             hash_md5.update(chunk)
     return hash_md5.hexdigest()
 
+
 def findIP():
     return ((([ip for ip in socket.gethostbyname_ex(socket.gethostname())[2] if not ip.startswith("127.")] or [[(s.connect(("8.8.8.8", 53)), s.getsockname()[0], s.close()) for s in [socket.socket(socket.AF_INET, socket.SOCK_DGRAM)]][0][1]]) + ["no IP found"])[0])
 
-from prompt_toolkit import prompt
-print('Flasher for Xiaomi Vacuum')
-print('trying to find the vacuum')
-
-
 
-
-process = subprocess.Popen(['mirobo','discover','--handshake','true'], stdout=subprocess.PIPE,stderr=subprocess.PIPE)
-time.sleep(1);
-process.terminate();
-output = process.stderr.read()
-
-
-#output = "IP 192.168.8.1: default - token: b'FFFFFFFFFFFFFFFFFFFFFF'"
-try:
-    m = re.search("b'(.+?)'", output)
-    token = m.group(1)
-    m = re.search("IP (.+?)[\: ]", output)
-    ip = m.group(1)
-except:
-    print('Vacuum not found...Check connection')
+def select_item(welcome_text, items):
+    print(welcome_text)
+    for i, item in enumerate(items):
+        print('{}. {}'.format(i+1, item))
+    try:
+        selected = input('Please select option by typing number (1-{}): '.format(len(items)))
+        result = items[int(selected)-1]
+        return result
+    except KeyboardInterrupt:
+        print('User requested to exit')
+        exit()
+    except ValueError:
+        print('Error! Please enter only one number')
+        exit()
+    except IndexError:
+        print('Error! Please enter one number between 1-{}'.format(len(items)))
+        exit()
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Flasher for Xiaomi Vacuum.\nFor specific options check \'{} --help\''.format(sys.argv[0]))
+    parser.add_argument('-a', '--address', dest='address', type=str, help='IP address of vacuum cleaner')
+    parser.add_argument('-t', '--token', dest='token', type=str, help='Known token of vacuum')
+    parser.add_argument('-f', '--firmware', dest='firmware', type=str, help='Path to firmware file')
+    parser.add_argument('-p', '--port', dest='http_port', type=str, default='8080', help='Port for local http server')
+
+    args, external = parser.parse_known_args()
+
+    print('Flasher for Xiaomi Vacuum')
+    print('Searching for device...')
+
+    ip_address = args.address
+    known_token = args.token
+    firmware = args.firmware
+
+    if not args.firmware:
+        print('You should specify firmware file name to install')
+        exit()
+
+    if not ip_address:
+        print('Address is not set. Trying to discover.')
```

Comment:
I think the python-miio offered a call which can be used to do the autodiscovery by the lib. It is `mirobot discover`

Reply:
which is not working as intended. Check https://github.com/rytilahti/python-miio/issues/152